### DISCUSSION TITLE
How can one use local variables in a closure (for/while loop)?

### DISCUSSION THREAD

**Discussion by anfelor** [koka-lang/koka#125] [Q&A]
Take this example from the Koka book, slightly modified:
```
fun fib2(n) {
  var x := 0
  var y := 1
  repeat(n) {
    var y0 := y
    y := x+y
    x := y0
  }
  x
}
```
Instead of declaring `y0` as a val, we declare it as a variable. I would expect this to work just as before, with `y0` as a local variable with initial value the same as `y` at that time. But instead I get a compiler error:
```
test.kk(43, 5): error: abstract type(s) escape(s) into the context
  term         :     var y0 := 2
                     y := x+y
  inferred type: () -> <local<$h>|_e> ()
  hint         : give a higher-rank type annotation to a function parameter?
```
I tried adding the type annotation (with `h` for `$h`) and using `local-scope`, but neither seemed to fix the problem.

**Comment by schicks** [comment]
So I think the problem is that `$h` is referring to a type that could be instantiated independently each time the closure is called. That is why the hint is to give a higher rank type annotation. What I'm not sure about is why the compiler thinks the local effect would ever exist with any type other than `int` in this context. The code below produces the same error, so the problem persists even if you remove all actual behavior from your example and add explicit annotations everywhere.
```
fun fib2(n: int): int {
  var x: int := 0
  var y: int := 1
  repeat(n) {
    var y0: int := 1
    y := 1
    x := 1
  }
  x
}
```

**Reply by anfelor** [reply]
I think `$h` is not `int` but rather a hidden Rank-2 type as in the [Haskell ST monad](http://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Monad-ST.html#v:runST) that guarantees that the local state doesn't "escape". I think the problem is not in the type of `y0` but that we have two local effects here (one for `x, y` and another for `y0`) and the compiler would need to do some unification to find out that they are the same and higher-order unification is hard.

**Comment by blakehawkins** [comment]
To state the obvious, you can circumvent this limitation by declaring `var` outside the looping scope and then simply re-initialise the var where you presently declare it

**Comment by TimWhiting** [comment]
```koka
fun fib2(n) {
  var x := 0
  var y := 1
  repeat(n) {
    var y0 := y
    y := x+y
    x := y0
  }
  x
}
```

So the problem with allowing this is that the closure passed to `repeat` could be stored by `repeat` into a global, or it could be returned by `repeat` and eventually escape the stack frame introduced by `fib2`. I know, in this case it doesn't, but you do have to worry about this case in general - especially with effects. If the closure gets passed to an effect operation, it really might not be obvious that it escapes the stack. 

To check if a closure escapes the stack, you need to do a control flow analysis (CFA). Control flow analysis for functional languages is an `O(n^3)` algorithm at minimum which really isn't something you really want to be doing during type checking. Once types are available, there are maybe some faster algorithms, but we are doing this in the middle of type checking, since it involves heap types. Additionally we aren't guaranteed the dependent definitions (like `repeat`) are already type checked.

My PhD research is looking at Demand Control Flow Analysis for functional languages, which is a recent technique for doing control flow analysis that can work on just a local area of a program, instead of needing to do a full control flow analysis of the whole program. Therefore you could just pay a bit of time for these localized issues, without paying the cost of doing escape analysis over the whole program. However, this means that type checking would be non-deterministic (if you refactor your code it could take longer and time out). Control flow analysis isn't typically used in type systems for this reason (in addition to the `O(n^3)` minimum time bound that is typical - especially since you want to type your whole program)! In this case we could maybe let the user specify the timeout or effort to spend, and even if you can't type it, you can still report an error and suggest using a `ref` instead. My research is looking more at using Demand CFA for optimizations, which can be best effort instead of type systems stuff. It's worth noting that Koka doesn't even know what some of the identifiers refer to until after type checking due to overloaded identifiers, so even if this were possible, we'd want to do it as a separate pass after type checking, which means delaying some type errors until that pass, and maybe having to adjust the heap types later on.

So in other words, the reason why this example cannot be typechecked is because stack allocated variables cannot be captured in heap allocated closures. If you want to pass it to a closure use a `ref` which is heap allocated, or dereference it prior to the capture in the closure so that it is a `val` not a `var`.

**Reply by timjs** [reply]
But, regardless of `y0` being declared as `var` or `val`, the closure passed to `repeat` still captures the stack allocated variables `x` and `y`, which could be problematic as you described it. Am I missing something?

**Reply by TimWhiting** [reply]
No, if it captures variables that are `val`s then it captures them by value (for value types) - which are stack allocated, or the object pointer (for ref types) - which are heap allocated. The binding is still immutable, and a copy is stored in the closure's environment. Because it is immutable the value in the closure's environment is guaranteed to be the same as the original binding. In contrast, vars are essentially a builtin algebraic effect that creates a location to store a value on the stack with one function to change what the value refers to (`:=`) and one function to reference the underlying value (using the variable name itself). Due to it being an effect where a handler is "installed" it is tied to the actual runtime stack location of the handler, and those functions to access and set the value need to be called under that stack location. Essentially it is precisely because it is a `var` instead of a `val`. I will note that the actual runtime implementation is more efficient than installing actual effect handler machinery, but it requires the same invariants, but guarantees that the location is stack allocated and doesn't escape. 

I will note that koka's type system it seems is able to somehow deal with this to some degree as long as you don't mix vars from two different scopes probably for the same reasons that having two handlers for the same effect always refers to the closest installed handler. The core libraries have a few examples of uses that work, and I'm going to try to understand this better myself.

If you are interacting with multiple variables from different scopes, it is just best to use `ref`s.

**Reply by TimWhiting** [reply]
Essentially, maybe all that is really needed is an `unmask`

```koka
fun fib2(n) {
  // initializes a local effect in the outer scope with vars x,y
  var x := 0
  var y := 1
  repeat(n) {
    // initializes / handles a local effect in the inner scope with vars y0
    var y0 := y
    // refer to the local effect in the outer scope
    mask<local>
      y := x+y
      // refer to the local effect in the closure scope
      x := unmask<local>{y0} 
  }
  x
}
```

Either that or use named effect handler machinery for local vars which would be much cleaner syntax wise, but probably not as optimizable.

**Reply by timjs** [reply]
Thanks for your lengthy reply. What I meant to ask is: what happens when the closure capturing `x` and `y` *`var`s* outlive the stackframe they are declared in?

**Reply by TimWhiting** [reply]
```koka
fun f()
  var x := 0
  fn()
    x + 1

fun main()
  f()()
```
```
abstract types do not match
context      :   var x
term         :   var x := 0
                 fn
inferred type: forall<h,e,e1> () -> <local<h>|e> (() -> <local<h>|e1> int)
expected type: forall<h1> () -> <local<h1>|_e> (() -> <local<$h>|_e1> int)
because      : a reference to a local variable escapes it's scope
```
As this program and error message demonstrates, Koka has actually made stack escaping part of the type system for local variables using effects. So I take back my statements that checking for stack escaping requires a control flow analysis (this would be the case if it wasn't part of the types). 

In general type systems have to be more restrictive, and prevent some valid programs to be typeable. 

In contrast control flow analysis can recover more complex invariants without restricting the kinds of valid programs you can write. 

So by making it part of the type system, Koka makes it so that you have to understand the type system and how it restricts the kinds of programs you write to ensure variables don't escape the stack. In this case you have to understand effect handlers.

The issue I believe is more related what I was trying to get across in the last message (the shadowing of local handlers). Essentially  a set of variables in the scope of a single function introduces a handler `local<h>` related to a specific scope/heap type `h`. That local effect can be omitted locally or propagated upwards, but it seems that it is even more restrictive than that, it cannot be propagated up (you cannot add `<local<h>>` as an effect unless it seems you have passed in a reference to a `local-var<h,int>` to relate it to a specific var). In other words, you cannot capture a local var in a closure that could escape the scope.

That being said, I don't think that the original issue is related to closure escaping, rather it is due to effect shadowing.

**Reply by timjs** [reply]
Thanks Tim. This is actually what I expected to happen.
